doc/protocol.rst

```diff
@@ -0,0 +1,270 @@
+content-type negotiation ([rfc7231 -Hypertext Transfer Protocol
+(HTTP/1.1): Semantics and Content]
+(http://tools.ietf.org/html/rfc7231)).  A request and responses
+maybe sent using either a single HTTP response body with
+content length or using HTTP 1.1 chunked transfer encoding.
+Servers are encouraged to implement chunked transfer encoding,
+as it efficiently support arbitrary length responses.  Clients
+must accept chunked transfer encoding, as required by HTTP 1.1.
+
+The HTTP 1.1 chunked transfer encoding error report model
+indicates failed transfers. A connection that is closed without
+receiving a trailer chunk must be treated as an error. When a
+trailer chunk is received, it must be interrogated to see if it
+contains a ``Status`` header that indicates an error.
+
+MIME type is in the form
+``application/ga4gh.v${apiversion}+${encoding}``. Where
+``${encoding}`` is the encoding format described below. The
+``${apiversion}`` is the dot-separate hierarchical API version of
+the GA4GH API. For requests, the ``Accepts`` header is used to
+specify the desired encoding and API version. The API version
+hierarchy requested can be as specific as required, with omitted
+minor versions resulting in the newest version available on the
+server matching. Thus, a version of ``1.5`` can be satisfied
+with version such as ``1.5``, or ``1.5.3``, etc. The individual
+dot versions are compared as integers.  The response
+``Content-Type`` contains the exact version of the API that was
+return.
+
+Data Object Layer
+-----------------
+
+The data object layer is implemented on top of the HTTP Internet
+layer.  Except for restrictions place on object alignment for
+checkpoint/resume (see below), the object layer view the
+Internet layer as a stream of bytes.  Data objects are encoded
+in this layer in either JSON or binary protocol buffer format.
+
+JSON ASCII message encoding
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Messages are encoding as UTF-8 text following the standard
+defined by Protocol Buffers V3 JSON encoding. While
+implementation are not required to use the Protocol Buffers
+software, the JSON encode must follow its
+specification. Messages are written to the stream in
+[line-delimited JSON format]
+(https://en.wikipedia.org/wiki/JSON_Streaming#Line_delimited_JSON).
+
+The MIME type for JSON encode is
+``application/ga4gh.v${apiversion}+JSON``.
+
+Binary message encoding
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Data objects may be encoded in an efficient binary format using
+Protocol Buffers V3 binary encoding format. Each message is
+preceded by a 32-bit byte length written in network byte order,
+followed by the message bytes.
+
+The MIME type for binary encode is
+``application/ga4gh.v${apiversion}+x-protobuf``.
+
+Transfer checkpoint/resume
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The GA4GH streaming protocol supports resume of a response by
+sending periodic checkpoint object that can be used to resume a
+transfer at that point. See the Rationale section for discussion
```

Comment:
The GA4GH streaming protocol supports resuming a response that was interrupted by sending periodic checkpoints where the transfer can rewind to and complete the transfer.